### ISSUE TITLE
关于FPS

### ISSUE THREAD

**Issue by zhuzhuzhu2** [chenjun2hao/DDRNet.pytorch#12]
大佬您好，我用了一块RTX 2080s测试了一下DDRNet-23slim的FPS，与论文中的数据相差非常大。想请问一下大佬没有没测试过FPS？
并且我得到的DDRNet-23slim的模型参数文件大小为23M，与论文中的5.7M相差也甚远，请问大佬知道这是怎么一回事吗？
非常期待您的回答

![image](https://user-images.githubusercontent.com/66657668/131103241-f1308475-ea74-4eaa-877c-b0906db93a3e.png)

**Comment by ydhongHIT** [comment]
> 大佬您好，我用了一块RTX 2080s测试了一下DDRNet-23slim的FPS，与论文中的数据相差非常大。想请问一下大佬没有没测试过FPS？
> 并且我得到的DDRNet-23slim的模型参数文件大小为23M，与论文中的5.7M相差也甚远，请问大佬知道这是怎么一回事吗？
> 非常期待您的回答
> 
> ![image](https://user-images.githubusercontent.com/66657668/131103241-f1308475-ea74-4eaa-877c-b0906db93a3e.png)

What about your FPS? Note that the model should be moved to GPU and excluding the bn (just set the bias of conv to 'True' and remove the bn for speed test). Just use the code in 'https://github.com/VITA-Group/FasterSeg/blob/master/tools/utils/darts_utils.py#L184'. The 2080Ti should be a desktop version.

**Comment by ydhongHIT** [comment]
> > 大佬您好，我用了一块RTX 2080s测试了一下DDRNet-23slim的FPS，与论文中的数据相差非常大。想请问一下大佬没有没测试过FPS？
> > 并且我得到的DDRNet-23slim的模型参数文件大小为23M，与论文中的5.7M相差也甚远，请问大佬知道这是怎么一回事吗？
> > 非常期待您的回答
> > ![image](https://user-images.githubusercontent.com/66657668/131103241-f1308475-ea74-4eaa-877c-b0906db93a3e.png)
> 
> What about your FPS? Note that the model should be moved to GPU and excluding the bn (just set the bias of conv to 'True' and remove the bn for speed test). Just use the code in 'https://github.com/VITA-Group/FasterSeg/blob/master/tools/utils/darts_utils.py#L184'.

**Comment by ydhongHIT** [comment]
> 大佬您好，我用了一块RTX 2080s测试了一下DDRNet-23slim的FPS，与论文中的数据相差非常大。想请问一下大佬没有没测试过FPS？
> 并且我得到的DDRNet-23slim的模型参数文件大小为23M，与论文中的5.7M相差也甚远，请问大佬知道这是怎么一回事吗？
> 非常期待您的回答
> 
> ![image](https://user-images.githubusercontent.com/66657668/131103241-f1308475-ea74-4eaa-877c-b0906db93a3e.png)

参数文件大小为23M是正常的，5.7M是数学意义上的网络参数数量。

**Comment by zhuzhuzhu2** [comment]
> > 大佬您好，我用了一块RTX 2080s测试了一下DDRNet-23slim的FPS，与论文中的数据相差非常大。想请问一下大佬没有没测试过FPS？
> > 并且我得到的DDRNet-23slim的模型参数文件大小为23M，与论文中的5.7M相差也甚远，请问大佬知道这是怎么一回事吗？
> > 非常期待您的回答
> > ![image](https://user-images.githubusercontent.com/66657668/131103241-f1308475-ea74-4eaa-877c-b0906db93a3e.png)
> 
> What about your FPS? Note that the model should be moved to GPU and excluding the bn (just set the bias of conv to 'True' and remove the bn for speed test). Just use the code in 'https://github.com/VITA-Group/FasterSeg/blob/master/tools/utils/darts_utils.py#L184'. The 2080Ti should be a desktop version.

Thanks very much for your reply.  
The FPS I mean is the prediction speed of DDRNet-23slim. And I tried to exclude the bn, but after I removed the bn, the prediction result changed to be very bad. So, do you mean I should remove all the bn in the model when I do the speed test?

**Comment by ydhongHIT** [comment]
> > > 大佬您好，我用了一块RTX 2080s测试了一下DDRNet-23slim的FPS，与论文中的数据相差非常大。想请问一下大佬没有没测试过FPS？
> > > 并且我得到的DDRNet-23slim的模型参数文件大小为23M，与论文中的5.7M相差也甚远，请问大佬知道这是怎么一回事吗？
> > > 非常期待您的回答
> > > ![image](https://user-images.githubusercontent.com/66657668/131103241-f1308475-ea74-4eaa-877c-b0906db93a3e.png)
> > 
> > 
> > What about your FPS? Note that the model should be moved to GPU and excluding the bn (just set the bias of conv to 'True' and remove the bn for speed test). Just use the code in 'https://github.com/VITA-Group/FasterSeg/blob/master/tools/utils/darts_utils.py#L184'. The 2080Ti should be a desktop version.
> 
> Thanks very much for your reply.
> The FPS I mean is the prediction speed of DDRNet-23slim. And I tried to exclude the bn, but after I removed the bn, the prediction result changed to be very bad. So, do you mean I should remove all the bn in the model when I do the speed test?

If you want to keep the prediction results, you can refer to https://github.com/pytorch/pytorch/blob/master/torch/nn/utils/fusion.py.  The 109 FPS is measured with an overclocked GPU but you can get 102 FPS with a normal 2080Ti. I will update the FPS in  the next version of the paper.

**Comment by ydhongHIT** [comment]
> > > 大佬您好，我用了一块RTX 2080s测试了一下DDRNet-23slim的FPS，与论文中的数据相差非常大。想请问一下大佬没有没测试过FPS？
> > > 并且我得到的DDRNet-23slim的模型参数文件大小为23M，与论文中的5.7M相差也甚远，请问大佬知道这是怎么一回事吗？
> > > 非常期待您的回答
> > > ![image](https://user-images.githubusercontent.com/66657668/131103241-f1308475-ea74-4eaa-877c-b0906db93a3e.png)
> > 
> > 
> > What about your FPS? Note that the model should be moved to GPU and excluding the bn (just set the bias of conv to 'True' and remove the bn for speed test). Just use the code in 'https://github.com/VITA-Group/FasterSeg/blob/master/tools/utils/darts_utils.py#L184'. The 2080Ti should be a desktop version.
> 
> Thanks very much for your reply.
> The FPS I mean is the prediction speed of DDRNet-23slim. And I tried to exclude the bn, but after I removed the bn, the prediction result changed to be very bad. So, do you mean I should remove all the bn in the model when I do the speed test?

Use tools in https://github.com/NVIDIA-AI-IOT/torch2trt to convert the model into tensorRT and you can achieve 135FPS with ddrnet_23_slim.

**Comment by zhuzhuzhu2** [comment]
Thanks very much. With your advice, I got the FPS closed to the FPS in the paper. DDRNet is really great network. Thanks for your kindly help.

**Comment by subake** [comment]
@ydhongHIT , @zhuzhuzhu2 , Please, can you provide more detailed instructions or maybe file example how to deactivate BatchNorm? I managed to fuse Conv and BatchNorm everywhere, except DAPPM and SegHead modules. I got a maximum of 80FPS with tensor size [1, 3, 1024, 2048]

**Comment by zhuzhuzhu2** [comment]
I am sorry that I used a different GPU. So I think my result maybe could not help you. But actually, I found the fusion of Conv and BatchNorm could only improve a little bit of the speed. Maybe my program of fusion is also not good.

**Comment by Bruce-Si** [comment]
> > 大佬您好，我用了一块RTX 2080s测试了一下DDRNet-23slim的FPS，与论文中的数据相差非常大。想请问一下大佬没有没测试过FPS？
> > 并且我得到的DDRNet-23slim的模型参数文件大小为23M，与论文中的5.7M相差也甚远，请问大佬知道这是怎么一回事吗？
> > 非常期待您的回答
> > ![image](https://user-images.githubusercontent.com/66657668/131103241-f1308475-ea74-4eaa-877c-b0906db93a3e.png)
> 
> What about your FPS? Note that the model should be moved to GPU and excluding the bn (just set the bias of conv to 'True' and remove the bn for speed test). Just use the code in 'https://github.com/VITA-Group/FasterSeg/blob/master/tools/utils/darts_utils.py#L184'. The 2080Ti should be a desktop version.

Thank you for your great work. I tried the code for speed test. The latency is about 10ms for a frame in gpu v100. That's great. But when I use `model(input)` for a real video, the latency ranges from 10ms to 200ms in the same environment, which is confusing... Have you met the same problem?
